libyul/backends/evm/SSAControlFlowGraph.cpp

```diff
@@ -0,0 +1,249 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include <libyul/backends/evm/SSAControlFlowGraph.h>
+
+#include <libsolutil/StringUtils.h>
+#include <libsolutil/Visitor.h>
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wtautological-compare"
+#include <fmt/ranges.h>
+#pragma GCC diagnostic pop
+
+#include <range/v3/view/zip.hpp>
+
+using namespace solidity;
+using namespace solidity::util;
+using namespace solidity::yul;
+
+namespace
+{
+class SSACFGPrinter
+{
+public:
+	SSACFGPrinter(SSACFG const& _cfg, SSACFG::BlockId _blockId)
+	{
+		printBlock(_cfg, _blockId, 0);
+	}
+	SSACFGPrinter(SSACFG const& _cfg, size_t _functionIndex, Scope::Function const& _function)
+	{
+		printFunction(_cfg, _functionIndex, _function);
+	}
+	friend std::ostream& operator<<(std::ostream& stream, SSACFGPrinter const& printer) {
+		stream << printer.m_result.str();
+		return stream;
+	}
+private:
+	static std::string varToString(SSACFG const& _cfg, SSACFG::ValueId _var) {
+		if (_var.value == std::numeric_limits<size_t>::max())
+			return "INVALID";
+		auto const& info = _cfg.valueInfo(_var);
+		return std::visit(
+			GenericVisitor{
+				[&](SSACFG::UnreachableValue const&) -> std::string {
+					return "[unreachable]";
+				},
+				[&](SSACFG::PhiValue const&) -> std::string {
+					return fmt::format("v{}", _var.value);
+				},
+				[&](SSACFG::VariableValue const&) -> std::string {
+					return fmt::format("v{}", _var.value);
+				},
+				[&](SSACFG::LiteralValue const& _literal) -> std::string {
+					return formatNumberReadable(_literal.value);
+				}
```

Comment:
We could simplify this to:
```suggestion
			GenericVisitor{
				[&](SSACFG::UnreachableValue const&) -> std::string {
					return "[unreachable]";
				},
				[&](SSACFG::LiteralValue const& _literal) -> std::string {
					return formatNumberReadable(_literal.value);
				},
				[&](auto const&) -> std::string {
					return fmt::format("v{}", _var.value);
				}				
```

Reply:
i'd rather keep the code duplication and get a compiler error if we add anything to the variant

Reply:
Indeed! Missed that ;)